File: scrape/taxi_trips/taxi_trips.go
package main

import (
	"database/sql"
	"encoding/json"
	"fmt"
	"log"
	"os"
	"time"

	"cloud.google.com/go/cloudsqlconn"
	"cloud.google.com/go/cloudsqlconn/postgres/pgxv4"
	"github.com/SebastiaanKlippert/go-soda"
	_ "github.com/lib/pq"
)

//https://data.cityofchicago.org/resource/wrvz-psew.json?$where=trip_miles > 0.1

//import API TOKEN

//create a struct to hold the data

type Trip []struct {
	TripID                  string `json:"trip_id"`
	TaxiID                  string `json:"taxi_id"`
	TripSeconds             string `json:"trip_seconds"`
	TripMiles               string `json:"trip_miles"`
	PickupCentroidLatitude  string `json:"pickup_centroid_latitude"`
	PickupCentroidLongitude string `json:"pickup_centroid_longitude"`
	DropoffCentroidLatitude  string `json:"dropoff_centroid_latitude"`
	DropoffCentroidLongitude string `json:"dropoff_centroid_longitude"`
	PickupCommunityArea     string `json:"pickup_community_area"`
	DropoffCommunityArea     string `json:"dropoff_community_area"`
}





func CSVSample(data_set string, query string) Trip {
	APP_TOKEN := "8nxBA5pgg7aPQ4fDbf4wj8BfM"
	url_format := fmt.Sprintf("https://data.cityofchicago.org/resource/%s", data_set)
	sodareq := soda.NewGetRequest(url_format, APP_TOKEN)
	sodareq.Format = "json"
	sodareq.Query.Where = query
	sodareq.Query.Limit = 1000



	// Gets the response
	//wait 1 second
	time.Sleep(1 * time.Second)
	resp, err := sodareq.Get()
	if err != nil {
		log.Fatal(err)
	}
	defer resp.Body.Close()

	// Decodes the response
	var trips Trip
	err = json.NewDecoder(resp.Body).Decode(&trips)
	if err != nil {
		log.Fatal(err)
	}

	return trips


}

func connect(connStr string) *sql.DB {
	cleanup, err:= pgxv4.RegisterDriver("cloudsql-postgres",cloudsqlconn.WithIAMAuthN())

	if err != nil {
		log.Fatal(err)
	}

	defer cleanup()

	db, err := sql.Open(
		"cloudsql-postgres",
		connStr)

	return db
}


func main() {
	// Get today's date
	//02/01/2023 12:00:00 AM
	//https://data.cityofchicago.org/api/id/wrvz-psew.json?$query=select *, :id order by `trip_start_timestamp` asc limit 100
	//Get everything after 2023-01-01
	
	where_statement := "trip_start_timestamp > '2022-01-01T00:00:00.000'"


	// Get a sample of taxi trips from the Socrata API from today
	trips := CSVSample("wrvz-psew", where_statement)

	host := os.Getenv("PGHOST")
    port := os.Getenv("PGPORT")
    user := os.Getenv("PGUSER")
    password := os.Getenv("PGPASS")
    dbname := "bronze"
    connStr := fmt.Sprintf("host=%s port=%s user=%s password=%s dbname=%s sslmode=disable", host, port, user, password, dbname)

	db := connect(connStr) // connect to the database

	stmt, err := db.Prepare(`INSERT INTO taxi_trips ( trip_id, taxi_id, trip_seconds, trip_miles, pickup_centroid_latitude, pickup_centroid_longitude, dropoff_centroid_latitude, dropoff_centroid_longitude, pickup_community_area, dropoff_community_area ) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10) ON CONFLICT (trip_id) DO NOTHING`)
	if err != nil {
	log.Fatal(err)
	}
	defer stmt.Close()


	// Insert each taxi trip into the 'taxi_trips' table using a prepared statement
	for _, trip := range trips {
		fmt.Println(trip.TripID)
		count := 0
		if trip.TripID != "" && trip.TripSeconds != "" && trip.TripMiles != "" && trip.PickupCommunityArea != "" && trip.DropoffCommunityArea != "" && trip.TaxiID != "" {
			_, err = stmt.Exec(			
				trip.TripID,
				trip.TaxiID,
				trip.TripSeconds,
				trip.TripMiles,
				trip.PickupCentroidLatitude,
				trip.PickupCentroidLongitude,
				trip.DropoffCentroidLatitude,
				trip.DropoffCentroidLongitude,
				trip.PickupCommunityArea,
				trip.DropoffCommunityArea,
				
			)
			if err != nil {
				log.Fatal(err)
			}
			fmt.Printf("Inserted trip with ID %s\n", trip.TripID)
			fmt.Printf("Count %d\n", count)
			count++
	}
}

}

File: src/taxi-trips/taxi-trips.py
import os 
import sqlalchemy
import pandas as pd
from shapely.geometry import Point, Polygon
import time
import psycopg2
import pandas as pd
import geojson
import json
import geopandas as gpd
import requests
user = 'postgres'
password = '432'
ip = '34.134.248.227'
db_name = 'bronze'
port = '5432'


# connect to database


count = 0
for x in range(300, 50000, 100):
    conn = psycopg2.connect(f'host={ip} dbname={db_name} user={user} password={password} port={port}')
    endpoint = f'https://data.cityofchicago.org/resource/wrvz-psew.json?$query=SELECT%0A%20%20%60trip_id%60%2C%0A%20%20%60taxi_id%60%2C%0A%20%20%60trip_start_timestamp%60%2C%0A%20%20%60trip_end_timestamp%60%2C%0A%20%20%60trip_seconds%60%2C%0A%20%20%60trip_miles%60%2C%0A%20%20%60pickup_census_tract%60%2C%0A%20%20%60dropoff_census_tract%60%2C%0A%20%20%60pickup_community_area%60%2C%0A%20%20%60dropoff_community_area%60%2C%0A%20%20%60fare%60%2C%0A%20%20%60tips%60%2C%0A%20%20%60tolls%60%2C%0A%20%20%60extras%60%2C%0A%20%20%60trip_total%60%2C%0A%20%20%60payment_type%60%2C%0A%20%20%60company%60%2C%0A%20%20%60pickup_centroid_latitude%60%2C%0A%20%20%60pickup_centroid_longitude%60%2C%0A%20%20%60pickup_centroid_location%60%2C%0A%20%20%60dropoff_centroid_latitude%60%2C%0A%20%20%60dropoff_centroid_longitude%60%2C%0A%20%20%60dropoff_centroid_location%60%2C%0A%20%20%60%3A%40computed_region_vrxf_vc4k%60%0AWHERE%20%60trip_seconds%60%20%3C%20{x}%0AORDER%20BY%20%60trip_start_timestamp%60%20DESC%20NULL%20LAST'

    # Pull data from the endpoint with a limit of 1000 rows

    response = requests.get(endpoint).json()
    print(response)

    #convert json to dataframe

    df = pd.DataFrame(response)

    # Recreate the table in the database



    sql_table = 'taxi_trips'


    # Create a cursor object
    cur = conn.cursor()

    # Create the table if it doesn't exist
    cur.execute(f'''CREATE TABLE IF NOT EXISTS {sql_table} (
                    trip_id VARCHAR(100),
                    trip_seconds INT,
                    trip_start_timestamp TIMESTAMP,
                    trip_miles FLOAT,
                    pickup_community_area INT,
                    dropoff_community_area INT,
                    pickup_centroid_latitude FLOAT,
                    pickup_centroid_longitude FLOAT,
                    dropoff_centroid_latitude FLOAT,
                    dropoff_centroid_longitude FLOAT
                )''')

    # Ensure the datatypes are correct


    # insert the data into the table
    for index, row in df.iterrows():
        # if there is a null value, skip the row
        if row.isnull().values.any():
            continue
        print(f'Inserting row {index} into {sql_table}')
        trip_id = row['trip_id']
        taxi_id = row['taxi_id']
        trip_seconds = row['trip_seconds']
        trip_start_timestamp = row['trip_start_timestamp']

        trip_miles = row['trip_miles']

        pickup_community_area = row['pickup_community_area']
        dropoff_community_area = row['dropoff_community_area']
        pickup_centroid_latitude = row['pickup_centroid_latitude']
        pickup_centroid_longitude = row['pickup_centroid_longitude']
        dropoff_centroid_latitude = row['dropoff_centroid_latitude']
        dropoff_centroid_longitude = row['dropoff_centroid_longitude']
        trip_id = str(trip_id)
        taxi_id = str(taxi_id)
        trip_seconds = int(trip_seconds)
        trip_start_timestamp = str(trip_start_timestamp)
        trip_miles = float(trip_miles)
        pickup_community_area = int(pickup_community_area)
        dropoff_community_area = int(dropoff_community_area)
        pickup_centroid_latitude = float(pickup_centroid_latitude)
        pickup_centroid_longitude = float(pickup_centroid_longitude)
        dropoff_centroid_latitude = float(dropoff_centroid_latitude)
        dropoff_centroid_longitude = float(dropoff_centroid_longitude)

        # insert the row
        cur.execute(f'INSERT INTO {sql_table} (trip_id, taxi_id, trip_seconds, trip_start_timestamp, trip_miles, pickup_community_area, dropoff_community_area, pickup_centroid_latitude, pickup_centroid_longitude, dropoff_centroid_latitude, dropoff_centroid_longitude) VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s) ON CONFLICT DO NOTHING', (trip_id, taxi_id, trip_seconds, trip_start_timestamp, trip_miles, pickup_community_area, dropoff_community_area, pickup_centroid_latitude, pickup_centroid_longitude, dropoff_centroid_latitude, dropoff_centroid_longitude))

    # Commit the changes
    conn.commit()

    # Close the cursor and the connection
cur.close()
conn.close()


